### PR TITLE
feat: handle S3 error responses with HTTP 200 status code

### LINKED ISSUE #199: S3 Customization to handle status code 200 Errors
Related to https://github.com/awslabs/smithy-kotlin/issues/323

Relevant information from @aajtodd copied from related ticket:

Here is what Go does for status code 200 Errors:

https://github.com/aws/aws-sdk-go-v2/blob/main/codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3ErrorWith200Status.java#L18
https://github.com/aws/aws-sdk-go-v2/blob/d05dd9b08d180c3696c33130548881966b56ea29/service/s3/internal/customizations/handle_200_error.go#L33

Once 200 errors are properly parsed, ensure that standard retry logic properly retries on `InternalError` as noted in the comment below.

### PR DESCRIPTION
**Pull request by aajtodd**
## Issue \#

fixes #199 

## Description of changes

Adds an interceptor that attempts to parse the response XML as an error response and resets the HTTP status code accordingly to trigger error handling.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: services/s3/common/test/aws/sdk/kotlin/services/s3/internal/Handle200ErrorsInterceptorTest.kt [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import aws.smithy.kotlin.runtime.http.HttpStatusCode
+import aws.smithy.kotlin.runtime.http.response.HttpResponse
+import aws.smithy.kotlin.runtime.httptest.buildTestConnection
+import aws.smithy.kotlin.runtime.util.Attributes
+import kotlinx.coroutines.test.runTest
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFailsWith
+
+class Handle200ErrorsInterceptorTest {
```
Comment (lauzadis) [inline services/s3/common/test/aws/sdk/kotlin/services/s3/internal/Handle200ErrorsInterceptorTest.kt:23]: You could also add a test ensuring error payloads are unmodified (tests L27 of your interceptor `if (!response.status.isSuccess()) return response`)

**Review by lauzadis — APPROVED** [review]

File: services/s3/common/test/aws/sdk/kotlin/services/s3/internal/Handle200ErrorsInterceptorTest.kt [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            httpClient = buildTestConnection {
+                expect(HttpResponse(HttpStatusCode.OK, body = HttpBody.fromBytes(payload)))
+            }
+        }
+
+    fun assertException(ex: S3Exception) {
+        val expectedMessage = "Please reduce your request rate."
+        assertEquals("SlowDown", ex.sdkErrorMetadata.errorCode)
+        assertEquals(expectedMessage, ex.sdkErrorMetadata.errorMessage)
+        assertEquals(expectedMessage, ex.sdkErrorMetadata.errorMessage)
```
Comment (ianbotsf) [inline services/s3/common/test/aws/sdk/kotlin/services/s3/internal/Handle200ErrorsInterceptorTest.kt:53]: **Nit**:

Double line.
Double line.

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/customization/s3/S3ErrorWith200StatusIntegration.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.shapes.ServiceShape
+
+/**
+ * Register interceptor to handle S3 error responses returned with an HTTP 200 status code.
+ * see [aws-sdk-kotlin#199](https://github.com/awslabs/aws-sdk-kotlin/issues/199)
+ */
+class S3ErrorWith200StatusIntegration : KotlinIntegration {
+    override fun enabledForService(model: Model, settings: KotlinSettings): Boolean =
+        model.expectShape<ServiceShape>(settings.service).isS3
```
Comment (ianbotsf) [inline codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/customization/s3/S3ErrorWith200StatusIntegration.kt:25]: **Question**: Should this apply to any other S3-adjacent services like S3 Control, Glacier, etc.?

**Review by ianbotsf — APPROVED** [review]

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/customization/s3/S3ErrorWith200StatusIntegration.kt (hunk 3 above)
Comment (aajtodd) [reply codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/customization/s3/S3ErrorWith200StatusIntegration.kt:25]: This is a good question, I actually think several of our customizations may need to apply to both s3 and s3 control. Not 100% on which ones though.

**Pull request merged**